cosmwasm/contracts/wormhole-ibc/src/contract.rs

```diff
@@ -0,0 +1,191 @@
+        govpacket.chain == Chain::from(state.chain_id),
+        format!(
+            "this governance VAA is for chain {}, which does not match this chain ({})",
+            u16::from(govpacket.chain),
+            state.chain_id
+        )
+    );
+
+    // governance VAA replay protection
+    if VAA_ARCHIVE.has(deps.storage, vaa.hash.as_slice()) {
+        bail!("governance vaa already executed");
+    }
+    VAA_ARCHIVE
+        .save(deps.storage, vaa.hash.as_slice(), &true)
+        .context("failed to save governance VAA to archive")?;
+
+    // match the governance action and execute the corresponding logic
+    match govpacket.action {
+        Action::UpdateChannelChain {
+            channel_id,
+            chain_id,
+        } => {
+            // validate that the chain_id for the channel is wormchain
+            // we should only be whitelisting IBC connections to wormchain
+            ensure!(
+                chain_id == Chain::Wormchain,
+                "whitelisted ibc channel not for wormchain"
+            );
+
+            let channel_id_str = String::from_utf8(channel_id.to_vec())
+                .context("failed to parse channel-id as utf-8")?;
+            let channel_id_trimmed = channel_id_str.trim_start_matches(char::from(0));
+
+            // update the whitelisted wormchain channel id
+            WORMCHAIN_CHANNEL_ID
+                .save(deps.storage, &channel_id_trimmed.to_string())
+                .context("failed to save channel chain")?;
+            Ok(Event::new("UpdateChannelChain")
+                .add_attribute("chain_id", chain_id.to_string())
+                .add_attribute("channel_id", channel_id_trimmed))
+        }
+    }
+}
+
+fn post_message_ibc(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: WormholeExecuteMsg,
+) -> anyhow::Result<Response> {
+    let channel_id = WORMCHAIN_CHANNEL_ID
+        .load(deps.storage)
+        .context("failed to load whitelisted wormchain channel id")?;
+
+    // compute the packet timeout
+    let packet_timeout = env.block.time.plus_seconds(PACKET_LIFETIME).into();
+
+    // compute the block height
+    let block_height = env.block.height.to_string();
+
+    // compute the transaction index
+    // (this is an optional since not all messages are executed as part of txns)
+    // (they may be executed part of the pre/post block handlers)
+    let tx_index = env.transaction.as_ref().map(|tx_info| tx_info.index);
+
+    // actually execute the postMessage call on the core contract
+    let mut res = core_execute(deps, env, info, msg).context("wormhole core execution failed")?;
+
+    res = match tx_index {
+        Some(index) => res.add_attribute("message.tx_index", index.to_string()),
+        None => res,
+    };
+    res = res.add_attribute("message.block_height", block_height);
+
+    // Send the result attributes over IBC on this channel
+    let packet = WormholeIbcPacketMsg::Publish {
+        msg: res.attributes.clone(),
```

Comment:
This clone looks unnecessary?

Reply:
Without it, `res` is partially moved. Since we return `res` at the end of the function and still want to emit the same attributes as part of the event, we need the clone here.